hazelcast/src/test/java/com/hazelcast/util/collection/OAHashSetBenchmark.java

```diff
@@ -0,0 +1,300 @@
+        private final int hash;
+        private final Integer value;
+        private final String valueAsStr;
+        private final String sqrtAsStr;
+        private final int equalsCostFactor;
+        private final int hashCodeCostFactor;
+
+        private ObjectWithExpensiveHashCodeAndEquals(int value, int equalsCostFactor, int hashCodeCostFactor) {
+            this.value = value;
+            this.valueAsStr = "Lorem ipsum dolor sit amet, consectetur adipiscing elit" + value;
+            this.sqrtAsStr = Double.toString(Math.sqrt(value));
+            this.equalsCostFactor = equalsCostFactor;
+            this.hashCodeCostFactor = hashCodeCostFactor;
+            this.hash = hashCodeInternal();
+        }
+
+        private int getHash() {
+            return hash;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            boolean equals = false;
+
+            for (int i = 0; i < equalsCostFactor; i++) {
+                equals = equalsInternal(o);
+            }
+
+            return equals;
+        }
+
+        private boolean equalsInternal(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            ObjectWithExpensiveHashCodeAndEquals that = (ObjectWithExpensiveHashCodeAndEquals) o;
+
+            if (value != null ? !value.equals(that.value) : that.value != null) {
+                return false;
+            }
+            if (valueAsStr != null ? !valueAsStr.equals(that.valueAsStr) : that.valueAsStr != null) {
+                return false;
+            }
+            return sqrtAsStr != null ? sqrtAsStr.equals(that.sqrtAsStr) : that.sqrtAsStr == null;
+        }
+
+        @Override
+        public int hashCode() {
+            int hashCode = 0;
+
+            for (int i = 0; i < hashCodeCostFactor; i++) {
+                hashCode = hashCodeInternal();
+            }
+
+            return hashCode;
+        }
+
+        private int hashCodeInternal() {
+            int result = value != null ? value.hashCode() : 0;
+            result = 31 * result + (valueAsStr != null ? valueAsStr.hashCode() : 0);
+            result = 31 * result + (sqrtAsStr != null ? sqrtAsStr.hashCode() : 0);
+            return result;
+        }
+    }
+
+    public static void main(String[] args) throws RunnerException {
+        Options opt = new OptionsBuilder()
+                .include(OAHashSetBenchmark.class.getSimpleName())
+                .resultFormat(ResultFormatType.JSON)
+                .addProfiler(GCProfiler.class)
+                //                .addProfiler(LinuxPerfProfiler.class)
+                //                .addProfiler(HotspotMemoryProfiler.class)
+                //                .shouldDoGC(true)
+                //                .verbosity(VerboseMode.SILENT)
+                .build();
+
+        new Runner(opt).run();
+    }
+
```

Comment:
Can you post some performance numbers? Interesting to see how well this data-structure performs. Would also be interesting to include gc profiler since a HashMap will generate more litter.

Reply:
Posted both.